### ISSUE TITLE
Custom button bound to swaylock causes clicking anywhere on waybar to suspend after resume

### ISSUE THREAD

**Issue by daboross** [Alexays/Waybar#991]
Hi! Bit of a weird issue here. I've added a custom button to my waybar config bound to `swaylock`. Works great, except after using it once, clicking anywhere on waybar locks the screen againl., clicking anywhere on waybar executes the same command again. Doesn't seem to matter where I click - empty space, other buttons, clicking anywhere locks the screen again.

My minimal config:

```json5
{
    "height": 30, // Waybar height (to be removed for auto height)
    "modules-left": ["sway/workspaces", "sway/mode"],
    "modules-center": ["sway/window"],
    "modules-right": ["custom/lock"],
    "custom/lock": {
        "format": "",
        "on-click": "swaylock -f -i ~/Pictures/Wallpapers/lockscreen.jpg",
    },
}
```
Doesn't need a style sheet to take effect.

My assumption is that there's some interaction between sway and waybar here where waybar doesn't get some input it needs after the click, and thus is left in an in-between state. I originally discovered this with a button which executes `systemctl suspend`, and reduced it to `swaylock` (since I have swaylock run on suspend in my sway config).

**Comment by abondis** [comment]
Hello, I seem to have a similar issue.

My command is "grimshot copy area", and here is what happens:
- any click on the bar will trigger grimshot, until I do a right click on the bar
- if I keep the left click pressed, then select my area with the right click (so right+left at the same time), grimshot works properly. Then I still need to do a right click on the bar to go back to normal behaviour of waybar.

this seems to be similar to #998 too.

**Comment by daboross** [comment]
If you don't already have a workaround for this, I've got one working for me. I'm not at all sure if all parts of this workaround are necessary, but this works:

```
"custom/suspend": {
    "format": "...",
    "one-click": "sh -c '(sleep 0.5s; systemctl suspend) & disown",
}
```

**Comment by abondis** [comment]
wow, thank you so much! I guessed there was something to do of that kind, but could not remember about 'disown'. Thank you!

**Comment by MuhammedTaha02** [comment]
> Hi! Bit of a weird issue here. I've added a custom button to my waybar config bound to `swaylock`. Works great, except after using it once, clicking anywhere on waybar locks the screen againl., clicking anywhere on waybar executes the same command again. Doesn't seem to matter where I click - empty space, other buttons, clicking anywhere locks the screen again.
> 
> My minimal config:
> 
> ```js
> {
>     "height": 30, // Waybar height (to be removed for auto height)
>     "modules-left": ["sway/workspaces", "sway/mode"],
>     "modules-center": ["sway/window"],
>     "modules-right": ["custom/lock"],
>     "custom/lock": {
>         "format": "",
>         "on-click": "swaylock -f -i ~/Pictures/Wallpapers/lockscreen.jpg",
>     },
> }
> ```
> 
> Doesn't need a style sheet to take effect.
> 
> My assumption is that there's some interaction between sway and waybar here where waybar doesn't get some input it needs after the click, and thus is left in an in-between state. I originally discovered this with a button which executes `systemctl suspend`, and reduced it to `swaylock` (since I have swaylock run on suspend in my sway config).

Is it possible to gather all session buttons under a single module?

**Comment by daboross** [comment]
> Is it possible to gather all session buttons under a single module?

Maybe? I'm not sure. This is just my custom config, so I haven't found any problem with having one module per button.

Do you think that would fix this issue?

**Comment by felixsanz** [comment]
it's still happening. did anyone found a fix for this?

**Comment by daboross** [comment]
> it's still happening. did anyone found a fix for this?

Have you tried the workaround in my comment above? https://github.com/Alexays/Waybar/issues/991#issuecomment-776315349

`sh -c '(sleep 0.5s; systemctl suspend)' & disown` has been working fine for me.

**Comment by felixsanz** [comment]
> > it's still happening. did anyone found a fix for this?
> 
> Have you tried the workaround in my comment above? [#991 (comment)](https://github.com/Alexays/Waybar/issues/991#issuecomment-776315349)
> 
> `sh -c '(sleep 0.5s; systemctl suspend)' & disown`, and similar with `swaylock`, have been working fine for me.

yes but didn't work for me. tried only sh -c, only disown, and both

**Comment by felixsanz** [comment]
@daboross oh! the key is the sleep! having this works!!

> "on-click": "sleep 0.5; waylock",

big thanks for making me try again!

**Comment by daboross** [comment]
Ah, huh, I'm glad! I hadn't gotten it to work without also doing `sh -c '...' & disown`. I'll have to try just that!

To clarify my intention with the workaround, the point of `& disown` is to cut off the child process so swaybar would stop waiting for it before "unclicking". But we also need to give it time to unclick, thus we need to sleep before locking.

Finally,  you can only "disown" a true subprocess. "(x && y)" is just two commands run in sequence (with the second only running if the first succeeds). `sh -c` wraps the sleep and swaylock together into a single subprocess, so it can be disowned.

**Comment by VAISHAKH-GK** [comment]
> 



> If you don't already have a workaround for this, I've got one working for me. I'm not at all sure if all parts of this workaround are necessary, but this works:
> 
> ```
> "custom/suspend": {
>     "format": "...",
>     "one-click": "sh -c '(sleep 0.5s; systemctl suspend)' & disown",
> }
> ```

Thank you this works,
But still this issue needed to be fixed.

**Comment by SolDoesTech** [comment]
Thank you so much!!! this was driving me bonkers trying to figure out why it wasn't working

**Comment by Luks17** [comment]
Can confirm that  `sleep` is the key to make it work. I don't understand why this is the case but if it works it works.

**Comment by creeperpaint** [comment]
So, has anyone found the reason this happens?

**Comment by daym** [comment]
> So, has anyone found the reason why this happens?

Not for sure but it could be the classic problem with either mouse capture ot being release when you lock or waybar losing the mouse-button-released event on the thing because swaylock runs before you can release the mouse button on the thing.

**Comment by Alexays** [comment]
Thanks for the detailed report. The root cause here isn't Waybar-side: when the on-click command (swaylock, a rofi/wofi menu, systemctl suspend, etc.) grabs the pointer or takes over the screen before you release the mouse button, the compositor never delivers the button-release event to Waybar, so the GTK button stays in its pressed state and re-fires on the next click anywhere on the bar. This is the same compositor pointer-grab behavior tracked upstream (e.g. hyprwm/Hyprland#1348), and several people confirmed a compositor upgrade fixed it.

Two clean fixes exist today:
- Use `on-click-release` instead of `on-click` (release-type actions are supported now), or
- Detach the command so Waybar isn't holding it, e.g. `sh -c '(sleep 0.2; swaylock -f) & disown'`.

Since the fix belongs in the compositor and there are supported workarounds, I'm closing this. Please reopen if you still hit it on a current sway/hyprland with `on-click-release`. 🙏
